### DISCUSSION TITLE
How to check whether the currently selected candidate is from a specific backend?

### DISCUSSION THREAD

**Discussion by kings2u** [company-mode/company-mode#1267] [Q&A] [answered]
I've noticed that if I select a candidate from the `company-yasnippet` backend using `company-complete-selection`, the default text for the first field of the Yasnippet snippet will unfortunately be deleted. I've also noticed that this issue can be avoided if I instead expand the snippet with `yas-expand`, e.g.,:
```
(define-key company-active-map (kbd "<return>") 'yas-expand)
```
I'm wanting to write a more specialized function so that RET will send `company-complete-selection` unless the currently selected candidate derives from the `company-yasnippet` backend, in which case RET will send `yas-expand`. How can I programmatically check whether the currently selected candidate is from the `company-yasnippet` backend?

Thank you for your time, as always.

**Comment by dgutov** [comment, marked as answer]
> I've noticed that if I select a candidate from the company-yasnippet backend using company-complete-selection, the default text for the first field of the Yasnippet snippet will unfortunately be deleted.

Do you have an example of such problematic snippet? Ideally point out one of the snippets from https://github.com/AndreaCrotti/yasnippet-snippets, since it's the common collection to have installed.

> How can I programmatically check whether the currently selected candidate is from the company-yasnippet backend?

If you know how to read the definition of `company-lighter`, you could adopt its logic.

**Reply by kings2u** [reply]
Sure, the "use-package-binding" snippet in emacs-lisp-mode is an example:
```
# -*- mode: snippet -*-
# contributor: Daniel Hitzel
# name: use-package binding
# key: upb
# --
:bind ("${1:binding}" . ${2:function-name})
```
Expanding the snippet is supposed to highlight the word "binding," giving the user an opportunity to type over it, but if you instead expand the snippet by typing "upb" and pressing `RET` on the snippet in the company candidates, the word "binding" in the first field gets deleted prematurely.

> If you know how to read the definition of company-lighter, you could adopt its logic.

I spent some time with the function and am afraid its lesson is beyond my grasp. 😅 I also looked at `company-diag` to see how it prints the used backend with no luck.

**Reply by dgutov** [reply]
> Expanding the snippet is supposed to highlight the word "binding," giving the user an opportunity to type over it, but if you instead expand the snippet by typing "upb" and pressing RET on the snippet in the company candidates, the word "binding" in the first field gets deleted prematurely.

Well, it does that here. Perhaps you are using some non-default Company configuration? Or are a disciple of Evil, for instance (just a guess, I don't know if it's a problem).

> I spent some time with the function and am afraid its lesson is beyond my grasp.

It's a variable. :wink: It is expanded by the function `format-mode-line`, see its docstring.

**Reply by kings2u** [reply]
> Perhaps you are using some non-default Company configuration?

You were right of course! I really need to discipline myself to test bugs with `emacs -Q` before asking for help. 

I still couldn’t figure how to adapt the logic of `company-lighter` to determine the backend of the currently selected candidate, but since my main problem has been solved, I can save that question for another time. I’ve bugged you enough for one day!
